// dvd-service-js-common/src/dvd-service-js-common.js
import ua from 'dvd-base-js-ua';
import login from 'dvd-service-js-login';
import Cookies from 'js-cookie';
import util from 'dvd-service-js-util';
import weixin from 'dvd-service-js-weixin';
import debug from 'dvd-service-js-debug';
import runtime from 'dvd-base-js-runtime';
import axios from 'axios';
import dvdServiceJsTj from 'dvd-service-js-tj';
// import $ from 'jquery';
// import fundebug from 'fundebug-javascript';
// import fundebug_node from 'fundebug-nodejs';

/**
 * 功能：检测是否需要去微信授权
 */

(function () {
  // Cookies.remove('is_auto_login', {
  //   domain: util.getBaseDomain()
  // });
  // alert(Cookies.get('is_auto_login'));

  // Cookies.remove('weixin_auth_try_times', {
  //   domain: util.getBaseDomain()
  // });
  // alert(Cookies.get('weixin_auth_try_times'));
  // return;

  // 微信授权短时间内尝试次数
  
  let weixin_auth_try_times = Cookies.get('weixin_auth_try_times');

  // 完成微信授权之后会在cookie设置is_auto_login=1,有这个标识了不需要再走授权逻辑
  if (ua.isWeiXin()
    && !login.isLogined()
    && Cookies.get('is_auto_login') === undefined
    && (weixin_auth_try_times === undefined || weixin_auth_try_times < 1)) {
    // if (ua.isWeiXin()) {

    // 设置尝试授权次数，每次失败1天以后重新尝试
    Cookies.set('weixin_auth_try_times', weixin_auth_try_times ? parseInt(weixin_auth_try_times) + 1 : 1, {
      domain: util.getBaseDomain(),
      path: '/',
      expires: 1,   // 有效时间1天
      // expires: 1 / 24 / 60    // 有效时间1分钟
    });

      weixin.goAuthPage(true, true, true, true);
  }
})();


// 检查是否开启debug控制台
debug.check();

/**
 * 功能：检测是否需要domain跳转
 */
function checkRedirect(domain) {
  // 如果不是店铺域名就不跳转
  if (location.host.split(".")[0].length < 4 && location.host.split(".")[0]!="sky") {
    console.log("非店铺域名，不跳转");
    return false;
  }

  // 当前域名与强制域名和小写强制域名都不符时，跳转到强制域名
  if (domain && domain !== location.hostname && ''.toLowerCase && new String(domain).toLowerCase() !== location.hostname) {
    // 替换当前域名
    location.replace(location.href.replace(location.hostname, domain));
    // 中断程序执行
    throw new Error(`即将跳转强制域名(${location.href})，已主动抛出异常中断当前页面js执行，请忽略此异常信息~`);
  }
}

/**
 * 功能：检测接口是否需要domain跳转
 * 说明：每个接口公参会返回强制跳转字段，如果这个字段有值并且与当前域名不同则跳转
 */
function checkRedirectByResponse(response) {
  if (response && response.code === '10010') {
    // 跳转强制domain
    location.replace(location.href.replace(location.hostname, `bravetime.${util.getBaseDomain()}`));

    // 中断程序执行
    throw new Error(`即将跳转强制域名(${location.href})，已主动抛出异常中断当前页面js执行，请忽略此异常信息~`);
  } else {
    checkRedirect(Cookies.get('force_domain'));

    // 接口response设置的cookie在部分手机上并不会立即生效
    setTimeout(function () {
      checkRedirect(Cookies.get('force_domain'));
    }, 100);
  }
}

/**
 * 功能：注册全局ajax事件，检测cookie是否需要强制跳转
 */
(function () {
  if (runtime.isClient()) {
    var $ = require('jquery');

    // jquery接口返回后
    $(document).ajaxComplete(function (event, xhr, settings) {
      checkRedirectByResponse(xhr.responseJSON);
    });

    // axios接口返回后
    axios.interceptors.response.use(function (response) {
      checkRedirectByResponse(response.data);
      return response;
    }, function (error) {
      checkRedirectByResponse();
      return Promise.reject(error);
    });
  }
})();


// 页面渲染完时关闭page-loading
(function () {
  let times = 0;
  // 关闭loading方法
  function close() {
    if (runtime.isClient() && times <= 0 && document.body.className.indexOf('loaded') === -1) {
      document.body.className += ' loaded';
      times++;
    }
  }

  // vue初始化1秒之后
  // if (window.Vue) {
  //   window.Vue.nextTick(function () {
  //     setTimeout(close, 1000);
  //   });
  // }

  if (runtime.isClient()) {
    var $ = require('jquery');
    // jquery接口返回后
    $(document).ajaxComplete(close);
  }

  // axios接口返回后
  axios.interceptors.response.use(function (response) {
    close();
    return response;
  }, function (error) {
    close();
    return Promise.reject(error);
  });

})();

/**
 * 错误统计
 */
(function () {
  if (runtime.isClient()) {
    let fundebug = require('fundebug-javascript');
    fundebug.apikey = "24d30241569618d701a631eca1e615b760cc1b7ba7a4f551a4ed0b9b89f393ae";
    fundebug.metaData = {cookie: document.cookie};
    fundebug.filters = [{
      req: {
        url: /api\.growingio\.com/
      }
    },
      {
        message: /^Script error\.$/
      },
      {
        fileName: /gio/
      }, {
        message: /WeixinJSBridge is not defined/
      }]
    fundebug.sampleRate = 0.01;
    fundebug.silentVideo = false;
  } else if (runtime.isServer()) {
    // fundebug_node.apikey = "f5b01856d83868dc14df3ed86033ba0a5bc1ec05007a5bd4fc26b3905d8b5819";
    // fundebug_node.sampleRate = 0.01;
  }

})();

(function(){
  dvdServiceJsTj.initAutoSelector();
})();

/**
 * @module dvd-service-js-common
 * @author swg [源码地址](http://gitlab.rd.vyohui.com/FE-Service/dvd-service-js-common.git)
 */
export default {
  checkRedirect: checkRedirectByResponse
}
