// dvd-service-js-tj/src/dvd-service-js-tj.js
import Cookie from 'js-cookie';
import login from 'dvd-service-js-login';
import ua from 'dvd-base-js-ua';
import runtime from 'dvd-base-js-runtime';
import ajax from 'dvd-service-js-ajax';
import getSelector from './chrome-similar-selector';
import $ from 'jquery';

// 前端点击数据采集格式，具体如下：
let defaultParam = {

  // ip
    ip: '',

    // ng时间，留空，数据由nginx生成
    nxtime: Date.now(),

    // 日志时间
    timestamp: Date.now(),

    // 业务线 默认为0，首页点击为1，
    production: 0,

    // 日志来源 默认为0，首页点击为1
    log_source: 0,

    // 浏览器UA
    user_agent: runtime.isClient() ? navigator.userAgent : '',

    // 来源市场，andriod是渠道号，ios为空
    market: '',

    // 用户id
    uid: login.getUserId(),

    // session id
    session: login.getSessionId(),

    // 卖家状态 (0：游客 1:买家 3:卖家)
    status: login.getUserStatusCode(),

    // 1：点击
    action: 1,

    // 1：模板
    action_type: 0,

    // 模板id
    object_id: '',

    // 设备号
    device_id: '',

    // 设备类型
    device: ua.isIos() ? 'ios' : ua.isAndroid() ? 'android' : '',

    // 设备版本号
    sys_version: '',

    // 分辨率
    resolution: runtime.isClient() ? window.screen.width + '*' + window.screen.height : 'SSR',

    // 当前位置
    location: runtime.isClient() ? location.href : 'https:bravetime.davdian.com',

    // APP版本号
    app_version: '',

    // 详细信息
    production_data: {
      feed: {

      // 整个feed item的位置，透传服务端下发的position
        itemPosition: '',

        // 模板Id
        tplId: '',

        // title or body
        type: '',

        // 当前点击的内容在body中的位置，透传服务端下发的position，title无此字段
        dataPosition: '1',

        // 动作：点击，来自feed中的command->content
        cmdContent: '',

        // 当前点击imgUrl，可以为空
        imgUrl: ''
      }
    }
  },

  // 点击统计默认数据
  defaultClickData = {
    resolution: runtime.isClient() ? window.screen.width + '*' + window.screen.height : 'SSR',
    token: Cookie.get('token'),
    platform: ua.isIos() ? 'ios_h5' : ua.isAndroid() ? 'android_h5' : '',
    host: runtime.isClient() ? location.host : 'bravetime.davdian.com',
    action: 'click'
  };

/**
 * @module dvd-service-js-tj
 * @author swg [源码地址](http://gitlab.rd.vyohui.com/FE-Service/dvd-service-js-tj.git)
 * @see http://wiki.bravetime.net/pages/viewpage.action?pageId=14778431
 */
export default {

  /**
   * 功能：发送统计信息
   * wiki :http://wiki.bravetime.net/pages/viewpage.action?pageId=14778431
   * 用法：
   * tj.send({
        production: 17,
        action: 1,
        action_type: action_type,
        production_data: {
          feed: {
            cmdContent: cmdContent || ''
          }
        }
      });
   */
  send(params) {
    let newParam = Object.assign({}, defaultParam, params);
    ajax.send({
      cache: false,
      async: false,

      // 这个接口值在davdian.com下存在
      url: '//bravetime.davdian.com/appapi?_=' + Date.now(),
      type: 'post',
      dataType: 'text',
      data: JSON.stringify(newParam),
      success(response) {
        console.log(response);
      },
      error(error) {
        console.error('ajax error:' + error.status + ' ' + error.statusText);
      }
    });
  },

  sendAllMeasure(page){
    var that = this;
    if(runtime.isClient()){
      if(window.performance){
        performance.getEntriesByType('mark').forEach(function(d){
          that.sendMeasure({
            page:page,
            name:d.name,
            measure:d.startTime
          })
        })
      }
    }
  },

  /**
    * 发送时间差
    * @param {String} params.page [当前页面]
    * @param {String} params.name [测量名称]
    * @param {Number} params.measure [测量时间]
    */
  sendMeasure(params){
    let platform = "";
    if(ua.isIos()){
      platform +="ios_";
    }else if(ua.isAndroid()){
      platform +="android_";
    }else{
      platform +="web_";
    }

    if(ua.isWeiXin()){
      platform +="wexin";
    }else if(ua.isDvdApp()){
      platform +="dvdapp";
    }else{
      platform +="other";
    }

    let data = {
      action:"measure",
      page:params.page,
      name:params.name,
      measure:params.measure|0,
      time: Date.now(),
      dvdsid: Cookie.get('dvdsid'),
      platform:platform,
    };
    $.ajax({
      cache:false,
      url:"https://tj.davdian.com/measure.php",
      type:"post",
      dataType:'text',
      data:JSON.stringify(data)
    });
  },

  /**
   * 初始化无埋点点击统计
   */
  initAutoSelector() {
    document.body.addEventListener('click', (e) => {
      let target, selector, data, newData;
      try {
        target = e.target;
        selector = getSelector(target);
        data = {
          action_data: selector,
          time: Date.now(),
          pos_x: e.pageX,
          pos_y: e.pageY,
          uri: runtime.isClient() ? window.dvd_tj_path || location.pathname : '',
          dvdsid: Cookie.get('dvdsid')
        };
        newData = Object.assign({}, defaultClickData, data);
        $.ajax({
          cache: false,
          async: false,
          url: 'https://tj.davdian.com/point.php',
          type: 'post',
          dataType: 'text',
          data: JSON.stringify(newData)
        });
      } catch (err) {
        console.error('initAutoSelector');
        console.error(err);
      }

    });
  }
};
