// dvd-base-js-page-scroll-position/dvd-base-js-page-scroll-position.js
import util from 'dvd-service-js-util';
import type from 'dvd-base-js-type';

/**
 * @module dvd-base-js-page-scroll-position
 * @author swg [源码地址](http://gitlab.rd.vyohui.com/FE-Base/dvd-base-js-page-scroll-position.git)
 */
export default {
  interval: null,
  /**
   * 根据（上次保存的页面滚动位置|给定的scrollTop）初始化页面滚动位置
   * @param saveType {Number} 0:sessionStorage，1:localStorage
   * @param key {String} 存入sessionStorage|localStorage的key
   * @param scrollTop {Number} 直接设置页面垂直滚动距离
   */
  init(saveType = 0, key = location.pathname + location.search + location.hash, scrollTop = null) {
    this.clearDisableBrowserLocationInterval();
    if (!type.isNumber(scrollTop)) {
      if (saveType === 0) {
        scrollTop = sessionStorage.getItem(key) || 0;
      } else if (saveType === 1) {
        scrollTop = localStorage.getItem(key) || 0;
      }
    }
    window.scrollTo(0, scrollTop);
  },
  /**
   * 自动保存页面滚动位置
   * @param saveType {Number} 0:sessionStorage，1:localStorage
   * @param key {String} 存入sessionStorage|localStorage的key
   */
  autoSave(saveType = 0, key = location.pathname + location.search + location.hash) {
    this.destory();
    this.callback = function (event) {
      let scrollTop = util.getDocumentScrollTop();
      if (saveType === 0) {
        sessionStorage.setItem(key, scrollTop);
      } else if (saveType === 1) {
        localStorage.setItem(key, scrollTop);
      }
    };
    window.addEventListener('scroll', this.callback, false);
  },
  /**
   * 解除事件绑定
   */
  destory() {
    window.removeEventListener('scroll', this.callback, false);
    this.setDisableBrowserLocationInterval();
  },
  /**
   * 设置定时器消除浏览器自带scrollTop定位
   */
  setDisableBrowserLocationInterval() {
    let ts = this;
    let times = 0;
    let scrollTop = util.getDocumentScrollTop();
    window.clearInterval(ts.interval);
    ts.interval = window.setInterval(function () {
      window.scrollTo(0, scrollTop);
      times++;
      // 一定次数后自动消除定时器
      if (times > 20) {
        window.clearInterval(ts.interval);
      }
    }, 10);
  },
  /**
   * 消除定时器消除浏览器自带scrollTop定位
   */
  clearDisableBrowserLocationInterval() {
    window.clearInterval(this.interval);
  },
};
