// dvd-base-build/dist/util.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _moment = require('moment');

var _moment2 = _interopRequireDefault(_moment);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _through = require('through2');

var _through2 = _interopRequireDefault(_through);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// 时间格式化
exports.default = {
  /* 获取当前格式化时间 */
  getNow: function getNow() {
    return (0, _moment2.default)().format('YYYY-MM-DD HH:mm:ss ' + (0, _moment2.default)().millisecond());
  },

  /* 获取被格式化当前时间作为静态资源版本号 */
  getTimeFormatVersion: function getTimeFormatVersion() {
    return (0, _moment2.default)().format("YYYY-MM-DD_HH:mm:ss");
  },

  /* 获取工程路径。注意：服务器端编译时取的是编译时的目录名，和本地不一致 */
  getProjectPath: function getProjectPath() {
    return _path2.default.resolve(__dirname, '../../../');
  },

  /* 获取工程名称。注意：服务器端编译时取的是编译时的目录名，和本地不一致 */
  getProjectName: function getProjectName() {
    return _path2.default.basename(this.getProjectPath());
  },

  // 对象格式化
  stringifyFormat: function stringifyFormat(obj) {
    return JSON.stringify(obj, ' ', 2);
  },

  /* gulp插件，什么也不做 */
  gulpNothing: function gulpNothing() {
    return _through2.default.obj(function (file, enc, cb) {
      this.push(file);
      cb();
    });
  }
};
