// dvd-service-js-encrypt/dvd-service-js-encrypt.js
import md5 from 'crypto-js/md5.js';
import param from 'dvd-base-js-param';
import ua from 'dvd-base-js-ua';
import login from 'dvd-service-js-login';

/**
 * @module dvd-service-js-encrypt
 * @author swg [源码地址](http://gitlab.rd.vyohui.com/FE-Service/dvd-service-js-encrypt.git)
 */
export default function (obj) {
  // 公共参数
  let data = {
    data_version: '0',
    device_token: '',
    format: 'json',
    osv: (function () {
      if (ua.isDvdApp()) {
        if (ua.isAndroid()) {
          return 'web_android_*_*';
        } else if (ua.isIos()) {
          return 'web_ios_*_*';
        }
      }
      return 'web_h5_*_*';
    })(),
    shop_url: `${location.protocol}//${location.host}`,
    sess_key: login.getDvdsid(),
    ts: Date.now(),
    wh: '750_1334',
    device_token: ua.getDvdAppDeviceId(),
  };

  // 统计参数
  let tjKeys = ['rp', 'rl', 'logDp', 'dp'];
  let url = /* 如果有VueRouter则用hask提取url */ window.VueRouter ? location.hash : location.href;
  for (let i in tjKeys) {
    let tjKey = tjKeys[i];
    let tjVal = param.get(tjKey, url);
    if (tjVal) {
      data[tjKey] = tjVal;
      // param[tjKey] = tjVal.replace(/[ +]/g, '');
    }
  }

  // 业务参数
  for (let i in obj) {
    data[i] = obj[i];
  }

  // map转map数组，编码
  let signArr = [];
  let dataStrArr = [];
  Object.keys(data).sort().forEach(function (key, i) {
    signArr.push(`${key}=${data[key]}`);
    dataStrArr.push(`${i === 0 ? '' : '&'}${key}=${encodeURIComponent(data[key])}`);
  });

  // md5校验值
  let sign = md5(signArr.join('')).toString().toUpperCase();

  return `${dataStrArr.join('')}&sign=${sign}`;
};
