// dvd-base-js-param/dist/dvd-base-js-param.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _dvdBaseJsType = require('dvd-base-js-type');

var _dvdBaseJsType2 = _interopRequireDefault(_dvdBaseJsType);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// url正则，用于获取url中各部分信息
var UrlReg = /^((?:http|https):)?(?:\/\/)?(([^:/]*)?(?::(\d+))?)?(\/[^?]*)?(\?[^#]*)?(#.*)?$/;

/**
 * @module dvd-base-js-param
 * @author sunweiguang [源码地址](http://gitlab.rd.vyohui.com/FE-Base/dvd-base-js-param.git)
 */
exports.default = {
  /**
   * 将url转成window.location对象
   * @param url {String}
   * @returns {{}} 如果传url参数，则返回与window.location相同格式的对象；如果不传url参数，则直接返回window.location
   */
  getLocation: function getLocation(url) {
    if (url) {
      var result = UrlReg.exec(url);
      return {
        href: result[0],
        protocol: result[1],
        host: result[2],
        hostname: result[3],
        port: result[4],
        pathname: result[5],
        search: result[6],
        hash: result[7]
      };
    } else {
      return window.location;
    }
  },


  /**
   * 将字符串转换为对象
   * @param searchStr {String} location.search格式的字符串，如:'?a=1&b=2&c=3'
   * @returns {{}} 比如{a:1,b:2,c:3}
   */
  toObj: function toObj(searchStr) {
    var obj = {};
    searchStr = searchStr || location.search.split('?')[1];
    if (searchStr) {
      var paramMapArr = searchStr.split("&");
      for (var i in paramMapArr) {
        var paramMap = paramMapArr[i].split("=");
        obj[paramMap[0]] = paramMap[1] || '';
      }
    }
    return obj;
  },


  /**
   * 将对象转换为字符串，与toObj方法相反
   * @param searchObj {{}} 比如{a:1,b:2,c:3}
   * @returns {String} location.search格式的字符串，如:'?a=1&b=2&c=3'
   */
  toStr: function toStr(searchObj) {
    var str = '';
    if (searchObj) {
      for (var i in searchObj) {
        str += '&' + i + '=' + searchObj[i];
      }
    }
    return str.substr(1);
  },


  /**
   * 获取url中的所有参数
   * @param url {String} 默认取当前页面url
   * @returns {{}}
   */
  getAll: function getAll(url) {
    var location = this.getLocation(url);
    var searchStr = location.search ? location.search.split("?")[1] : location.search;
    return this.toObj(searchStr);
  },


  /**
   * 获取url中的某个参数
   * @param name {String} 参数名称
   * @param url {String} 默认取当前页面url
   * @returns {*}
   */
  get: function get(name, url) {
    if (name) {
      return this.getAll(url)[name];
    } else {
      throw new Error('name参数不能为空');
    }
  },


  /**
   * 替换url中的某个参数，如果不存在则添加
   * @param map {{}} 包含需要替换的key/value
   * @param url {String} 被替换的url，默认取当前页面url
   * @returns {string}
   */
  replace: function replace(map, url) {
    url = url || location.href;
    if (map) {
      var obj = this.getAll(url);
      for (var i in map) {
        obj[i] = map[i];
      }
      return url.split("?")[0] + '?' + this.toStr(obj);
    } else {
      throw new Error('map参数不能为空');
    }
  },


  /**
   * 向url中添加参数，如果已存在则修改
   * @param map {{}} 包含需要添加的key/value
   * @param url {String} 被操作的url，默认取当前页面url
   * @returns {string}
   */
  add: function add(map, url) {
    return this.replace(map, url);
  },


  /**
   * 移除url中参数
   * @param key {String|Array} 删除的参数名，可以是字符串或者字符串数组
   * @param url {String} 被操作的url，默认取当前页面url
   * @returns {string}
   */
  remove: function remove(key, url) {
    url = url || location.href;
    if (key) {
      var obj = this.getAll(url);
      if (_dvdBaseJsType2.default.isString(key)) {
        delete obj[key];
      } else if (_dvdBaseJsType2.default.isArray(key)) {
        for (var i in key) {
          delete obj[key[i]];
        }
      }
      return url.split("?")[0] + '?' + this.toStr(obj);
    }
  }
};
