// dvd-service-js-debug/dvd-service-js-debug.js
import cookie from 'js-cookie';
import scriptjs from 'scriptjs';
import util from 'dvd-service-js-util';
import popup from 'dvd-service-js-popup';

const config = {
  threshold: 10, // 阈值：计步器达到阈值时开启开发者模式
  step: 0, // 计步器：达到阈值时开启开发者模式
  cookieFlag: {
    developer: 'h5_developer_mode_flag', // 开发者模式开关
    debug: 'h5_debug_flag', // 调试控制台开关
  },
  developerPageUrl: '/developer.html', // 开发者页面地址
};

/**
 * @module dvd-service-js-debug
 * @author swg [源码地址](http://gitlab.rd.vyohui.com/FE-Service/dvd-service-js-debug.git)
 */
export default {
  /**
   * 检测调试状态，自动开启某些功能
   */
  check() {
    // 检测调试开关状态。如果开关打开，则在当前页面立即打开vConsole。
    if (this.checkDebugFlag()) {
      this.openDebug();
    }
  },

  /**
   * 检测-是否开启开发者模式
   * @returns {Boolean}
   */
  checkDeveloperFlag() {
    return cookie.get(config.cookieFlag.developer);
  },
  /**
   * 进入-开发者页面
   */
  goDeveloperPage() {
    let ts = this;

    // 如果已经是开发者模式，直接进入开发者页面
    if (cookie.get(config.cookieFlag.developer)) {
      ts.openDeveloper();
    } else {
      // 计步器+1
      let lastStep = ++config.step;
      console.log(`调试开关计步器当前为${config.step}`);

      // 如果达到门槛，输入密码正确后，进入开发者页面
      if (config.step >= config.threshold) {
        // 开启调试工具
        popup.prompt({
          className: 'debug',
          title: '即将启动H5开发者模式',
          text: ' ',
          placeholder: '请输入密码',
          btnTitle: '',
          okBtnCallback(result) {
            if (result == 'dvd@1ppt') {
              ts.openDeveloper();
            } else {
              popup.toast('密码错误，请重新输入', 2000, 'debug');
              return false;
            }
          },
          cancelBtnTitle: '',
          cancelBtnCallback() {
          }
        });

        // 重置计步器
        config.step = 0;
        // 如果以上2种情况都说不是，点击次数+1
      } else {
        // 如果没有连续调用，则计步器重置为0
        setTimeout(function () {
          if (config.step === lastStep) {
            config.step = 0;
            console.log('调试开关计步器已重置为0');
          }
        }, 500);
      }
    }
  },
  /**
   * 开启-开发者模式
   */
  openDeveloper(param = {}) {
    // 自动设置调试开关
    if (!this.checkDeveloperFlag()) {
      cookie.set(config.cookieFlag.developer, 1, {domain: util.getBaseDomain(), expires: 365, path: '/'});
    }

    // 跳转开发者模式页
    location.href = config.developerPageUrl;
  },
  /**
   * 关闭-开发者模式
   */
  closeDeveloper() {
    cookie.remove(config.cookieFlag.developer, {domain: util.getBaseDomain(), path: '/'});
    location.reload();
  },

  /**
   * 检测-是否开启调试控制台
   * @returns {Boolean}
   */
  checkDebugFlag() {
    return cookie.get(config.cookieFlag.debug);
  },
  /**
   * 开启-调试控制台
   */
  openDebug() {
    // vConsole需要手机支持Symbol类型
    if (window.Symbol) {
      let VConsole = require('vconsole');
      // 自动设置调试开关
      if (!this.checkDebugFlag()) {
        cookie.set(config.cookieFlag.debug, 1, {domain: util.getBaseDomain(), expires: 365, path: '/'});
      }

      // 大V红
      let style = document.createElement('style');
      style.type = 'text/css';
      style.innerHTML = '#__vconsole .vc-switch { background-color: #ff4a7d !important; right: 10px !important; bottom: 60px !important;}';
      document.head.appendChild(style);

      // 初始化VConsole调试工具
      /*if (VConsole) {
        VConsole.default ? new VConsole.default() : new VConsole();
      } else {
        scriptjs('//dcn-ws.davdian.com/vconsole/3.0.0/vconsole.min.js', function () {
          new window.VConsole();
        });
      }*/
      new VConsole();
    } else {
      alert('对不起，您的手机系统版本过低，不支持vConsole。');
    }
  },
  /**
   * 关闭-调试控制台
   */
  closeDebug() {
    cookie.remove(config.cookieFlag.debug, {domain: util.getBaseDomain(), path: '/'});
    location.reload();
  },
};
