// dvd-service-js-tj/dist/dvd-service-js-tj.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _jsCookie = require('js-cookie');

var _jsCookie2 = _interopRequireDefault(_jsCookie);

var _dvdServiceJsLogin = require('dvd-service-js-login');

var _dvdServiceJsLogin2 = _interopRequireDefault(_dvdServiceJsLogin);

var _dvdBaseJsUa = require('dvd-base-js-ua');

var _dvdBaseJsUa2 = _interopRequireDefault(_dvdBaseJsUa);

var _dvdBaseJsRuntime = require('dvd-base-js-runtime');

var _dvdBaseJsRuntime2 = _interopRequireDefault(_dvdBaseJsRuntime);

var _dvdServiceJsAjax = require('dvd-service-js-ajax');

var _dvdServiceJsAjax2 = _interopRequireDefault(_dvdServiceJsAjax);

var _chromeSimilarSelector = require('./chrome-similar-selector');

var _chromeSimilarSelector2 = _interopRequireDefault(_chromeSimilarSelector);

var _jquery = require('jquery');

var _jquery2 = _interopRequireDefault(_jquery);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// 前端点击数据采集格式，具体如下：
var defaultParam = {

  // ip
  ip: '',

  // ng时间，留空，数据由nginx生成
  nxtime: Date.now(),

  // 日志时间
  timestamp: Date.now(),

  // 业务线 默认为0，首页点击为1，
  production: 0,

  // 日志来源 默认为0，首页点击为1
  log_source: 0,

  // 浏览器UA
  user_agent: _dvdBaseJsRuntime2.default.isClient() ? navigator.userAgent : '',

  // 来源市场，andriod是渠道号，ios为空
  market: '',

  // 用户id
  uid: _dvdServiceJsLogin2.default.getUserId(),

  // session id
  session: _dvdServiceJsLogin2.default.getSessionId(),

  // 卖家状态 (0：游客 1:买家 3:卖家)
  status: _dvdServiceJsLogin2.default.getUserStatusCode(),

  // 1：点击
  action: 1,

  // 1：模板
  action_type: 0,

  // 模板id
  object_id: '',

  // 设备号
  device_id: '',

  // 设备类型
  device: _dvdBaseJsUa2.default.isIos() ? 'ios' : _dvdBaseJsUa2.default.isAndroid() ? 'android' : '',

  // 设备版本号
  sys_version: '',

  // 分辨率
  resolution: _dvdBaseJsRuntime2.default.isClient() ? window.screen.width + '*' + window.screen.height : 'SSR',

  // 当前位置
  location: _dvdBaseJsRuntime2.default.isClient() ? location.href : 'https:bravetime.davdian.com',

  // APP版本号
  app_version: '',

  // 详细信息
  production_data: {
    feed: {

      // 整个feed item的位置，透传服务端下发的position
      itemPosition: '',

      // 模板Id
      tplId: '',

      // title or body
      type: '',

      // 当前点击的内容在body中的位置，透传服务端下发的position，title无此字段
      dataPosition: '1',

      // 动作：点击，来自feed中的command->content
      cmdContent: '',

      // 当前点击imgUrl，可以为空
      imgUrl: ''
    }
  }
},


// 点击统计默认数据
defaultClickData = {
  resolution: _dvdBaseJsRuntime2.default.isClient() ? window.screen.width + '*' + window.screen.height : 'SSR',
  token: _jsCookie2.default.get('token'),
  platform: _dvdBaseJsUa2.default.isIos() ? 'ios_h5' : _dvdBaseJsUa2.default.isAndroid() ? 'android_h5' : '',
  host: _dvdBaseJsRuntime2.default.isClient() ? location.host : 'bravetime.davdian.com',
  action: 'click'
};

/**
 * @module dvd-service-js-tj
 * @author swg [源码地址](http://gitlab.rd.vyohui.com/FE-Service/dvd-service-js-tj.git)
 * @see http://wiki.bravetime.net/pages/viewpage.action?pageId=14778431
 */
exports.default = {

  /**
   * 功能：发送统计信息
   * wiki :http://wiki.bravetime.net/pages/viewpage.action?pageId=14778431
   * 用法：
   * tj.send({
        production: 17,
        action: 1,
        action_type: action_type,
        production_data: {
          feed: {
            cmdContent: cmdContent || ''
          }
        }
      });
   */
  send: function send(params) {
    var newParam = _extends({}, defaultParam, params);
    _dvdServiceJsAjax2.default.send({
      cache: false,
      async: false,

      // 这个接口值在davdian.com下存在
      url: '//bravetime.davdian.com/appapi?_=' + Date.now(),
      type: 'post',
      dataType: 'text',
      data: JSON.stringify(newParam),
      success: function success(response) {
        console.log(response);
      },
      error: function error(_error) {
        console.error('ajax error:' + _error.status + ' ' + _error.statusText);
      }
    });
  },
  sendAllMeasure: function sendAllMeasure(page) {
    var that = this;
    if (_dvdBaseJsRuntime2.default.isClient()) {
      if (window.performance) {
        performance.getEntriesByType('mark').forEach(function (d) {
          that.sendMeasure({
            page: page,
            name: d.name,
            measure: d.startTime
          });
        });
      }
    }
  },


  /**
    * 发送时间差
    * @param {String} params.page [当前页面]
    * @param {String} params.name [测量名称]
    * @param {Number} params.measure [测量时间]
    */
  sendMeasure: function sendMeasure(params) {
    var platform = "";
    if (_dvdBaseJsUa2.default.isIos()) {
      platform += "ios_";
    } else if (_dvdBaseJsUa2.default.isAndroid()) {
      platform += "android_";
    } else {
      platform += "web_";
    }

    if (_dvdBaseJsUa2.default.isWeiXin()) {
      platform += "wexin";
    } else if (_dvdBaseJsUa2.default.isDvdApp()) {
      platform += "dvdapp";
    } else {
      platform += "other";
    }

    var data = {
      action: "measure",
      page: params.page,
      name: params.name,
      measure: params.measure | 0,
      time: Date.now(),
      dvdsid: _jsCookie2.default.get('dvdsid'),
      platform: platform
    };
    _jquery2.default.ajax({
      cache: false,
      url: "https://tj.davdian.com/measure.php",
      type: "post",
      dataType: 'text',
      data: JSON.stringify(data)
    });
  },


  /**
   * 初始化无埋点点击统计
   */
  initAutoSelector: function initAutoSelector() {
    document.body.addEventListener('click', function (e) {
      var target = void 0,
          selector = void 0,
          data = void 0,
          newData = void 0;
      try {
        target = e.target;
        selector = (0, _chromeSimilarSelector2.default)(target);
        data = {
          action_data: selector,
          time: Date.now(),
          pos_x: e.pageX,
          pos_y: e.pageY,
          uri: _dvdBaseJsRuntime2.default.isClient() ? window.dvd_tj_path || location.pathname : '',
          dvdsid: _jsCookie2.default.get('dvdsid')
        };
        newData = _extends({}, defaultClickData, data);
        _jquery2.default.ajax({
          cache: false,
          async: false,
          url: 'https://tj.davdian.com/point.php',
          type: 'post',
          dataType: 'text',
          data: JSON.stringify(newData)
        });
      } catch (err) {
        console.error('initAutoSelector');
        console.error(err);
      }
    });
  }
};
