// dvd-service-js-share/dvd-service-js-share.js
import ua from 'dvd-base-js-ua';
import native from 'dvd-service-js-native';
import weixin from 'dvd-service-js-weixin';
import Vue from 'vue';

let getEl = function () {
  return (document.querySelector('.app') || document.body).appendChild(document.createElement('div'));
};

/**
 * @module dvd-service-js-share
 * @author swg [源码地址](http://gitlab.rd.vyohui.com/FE-Service/dvd-service-js-share.git)
 */
export default {
  /**
   * 设置 微信&native 分享信息
   * 调用方法:
   share.setShareInfo({
      title: '大V店组团包邮', // 分享标题
      desc: '一件包邮！每天上新！好货低价又包邮，抢到了就赚翻啦', // 分享描述
      link: location.href, // 分享链接
      imgUrl: 'http://pic.davdian.com/free/2016/04/09/320_320_0fc3e0dbbadd249b7f1b93a525f0adf0.jpg', // 分享图标
    });
   跟app协商了 以后图文链接分享图片大小规定200X200 可以避免在app里分享不了的问题 oss上的图片可以加样式 ?x-oss-process=image/resize,m_lfit,h_200,w_200
   */
  setShareInfo(param = {}, response) {
    if (ua.isWeiXin()) {
      weixin.setShareInfo(param, response);
    } else if (ua.isDvdApp()) {
      native.custom.setShareInfo(param);
    }
  },
  /**
   * 唤起浏览器分享(目前只支持弹出浮层引导分享)
   */
  callBrowserShare() {
    new Vue({
      components: {
        'com-share-pop-tip': require('dvd-service-com-web-share-guide').default,
      },
      el: getEl(),
      data: {},
      template: '<com-share-pop-tip/>',
    });
  },
  /**
   * 唤起 浏览器|native 分享
   */
  callShare() {
    if (ua.isDvdApp()) {
      native.custom.share();
    } else {
      this.callBrowserShare();
    }
  },
};
